File: Estruturas de Dados/Avaliações/Trabalho02/principal_clima.py
import csv

def main():
    lista = []
    nome_base = "base.csv"

    try:
        # Abre o arquivo com o gerenciador de contexto 'with' (igual ao try-with-resources)
        with open(nome_base, mode='r', encoding='utf-8') as arquivo:
            leitor = csv.reader(arquivo)
            
            for dados_linha in leitor:
                # No Python, se o CSV tiver espaços após a vírgula, 
                # pode ser necessário usar dados_linha[0].strip()
                obj_clima = Clima(dados_linha[0], dados_linha[1], dados_linha[2], dados_linha[3])

                # O operador 'in' usa o método __eq__ definido na classe
                if obj_clima not in lista:
                    lista.append(obj_clima)
    

        # Exibe os objetos da lista
        for item in lista:
            print(item)

        # --- QUAIS MESES SÃO MAIS QUENTES? ---
        
        meses_mais_quentes = []
        maior_valor = float('-inf') # Equivalente ao Integer.MIN_VALUE

        for item in lista:
            valor_temp = item.get_valor_temperatura()

            if valor_temp > maior_valor:
                maior_valor = valor_temp
                meses_mais_quentes = [item] # Limpa e adiciona em uma única linha
            elif valor_temp == maior_valor:
                meses_mais_quentes.append(item)

        # Verifica e exibe resultados
        if meses_mais_quentes:
            print("\nMeses mais quentes:\n")
            for clima in meses_mais_quentes:
                print(clima)
        else:
            print("\nNenhum mês encontrado.")

    except FileNotFoundError:
        print(f"Erro: O arquivo '{nome_base}' não foi encontrado.")
    except Exception as e:
        print(f"Ocorreu algum erro... {e}")

if __name__ == "__main__":
    main()

File: Estruturas de Dados/Avaliações/Trabalho02/PrincipalClima.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class PrincipalClima {
    public static void main(String[] args) {

        //lista para armazenar os objetos da classe Clima
        List<Clima> lista = new ArrayList<>();

        // nome do arquivo a ser lido
        String nomeBase = "base.csv";

        // O try-with-resources já fecha o arquivo automaticamente (igual ao 'with' do Python)
        //abre o arquivo para leitura
        try (BufferedReader leitor = new BufferedReader(
                new FileReader(nomeBase, StandardCharsets.UTF_8))) {
            
            String linha;

            //Lê o arquivo linha por linha
            while ((linha = leitor.readLine()) != null) {

                //divide a linha usando vírgula
                String[] dadosLinha = linha.split(",");
                
                //cria um objeto da classe Clima usando os dados da linha
                Clima objClima = new Clima(dadosLinha[0],dadosLinha[1],dadosLinha[2], dadosLinha[3]);

                // Requer que o método equals() esteja implementado na classe Clima
                //evita a adição de objetos duplicados na lista
                if (!lista.contains(objClima)) {
                    lista.add(objClima);
                }
            }

            //exibe os objetos da lista
            for (Clima item : lista) {
                System.out.println(item);
            }

            //  QUAIS MESES SÃO MAIS QUENTES?

            //lista para armazenar os meses mais quentes encontrados
            List<Clima> mesesMaisQuentes = new ArrayList<>();

            //começa com o menor valor possível para garantir que qualquer temperatura encontrada seja maior
            int maiorValor = Integer.MIN_VALUE;

            //percorre toda a lista
            for (Clima item : lista) {
                
                //converte a temperatura do mês atual para um valor numérico
                int valorTemp = item.getValorTemperatura();

                //compara o valor da temperatura atual com o maior valor encontrado até agora
                if (valorTemp > maiorValor) {
                    maiorValor = valorTemp; //atualiza o maior valor encontrado
                    mesesMaisQuentes.clear(); // Limpa a lista para armazenar apenas os meses mais quentes
                    mesesMaisQuentes.add(item); // Adiciona o mês mais quente encontrado
                } else if (valorTemp == maiorValor) {
                    mesesMaisQuentes.add(item); // Adiciona outros meses que tenham a mesma temperatura mais alta
                }
            }

            //verifica se foram encontrados meses mais quentes e exibe os resultados
            if(!mesesMaisQuentes.isEmpty()) {
                System.out.println("\nMeses mais quentes:\n");

                //mostra os meses mais quentes encontrados
                for (Clima clima : mesesMaisQuentes) {
                    System.out.println(clima);
                }
            } else {

                //caso nenhum mês seja encontrado
                System.out.println("\nNenhum mês encontrado.");

            }

            // System.out.println("\nO mês mais quente é: " + maisQuente.mes + " do ano " + maisQuente.ano);
        
        } catch (Exception e) {

            //caso ocorra algum erro durante a leitura do arquivo
            System.err.println("Ocorreu algum erro... " + e.getMessage());
        }
    }
}
